cmd/mdatagen/internal/samplereceiver/internal/metadata/telemetry_test.go

```diff
@@ -0,0 +1,96 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package metadata
+
+import (
+	"context"
+	"testing"
+
+	io_prometheus_client "github.com/prometheus/client_model/go"
+	"github.com/stretchr/testify/require"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/component/componenttest"
+)
+
+// TODO: The goal of this functionality is to have it be generated by mdatagen. This is meant to be an example
+// of what the output of mdatagen should be. This method will be expanded when generated by mdatagen to run
+// tests for each telemetry metric.
+func TestTelemetryMetrics(t *testing.T) {
+	// Current status: Shows how to properly test gauge, sum, histogram, optional, and async metrics.
+	// Also tests with and without attributes on a telemetry metric.
+	// This is not a comprehensive test of the sample receiver's telemetry metrics, but should cover each of its cases.
+
+	// Component telemetry setup should be uniform across components
+	tt, err := componenttest.SetupTelemetry(component.MustNewID(Type.String()))
```

Comment:
is it necessary to test these using `componenttest` or can they be tested using the code in `generated_component_telemetry_test.go`?

Reply:
See https://github.com/open-telemetry/opentelemetry-collector/blob/main/cmd/mdatagen/internal/samplereceiver/generated_component_telemetry_test.go as an example

Reply:
It looks like it's possible to use that functionality, but there would be some level of impact.

A couple of things to call out:
1. The generated telemetry test from this PR would need to be moved up to each component's main directory, rather than `internal/metadata`. We could potentially append this testing to to the `generated_component_telemetry_test.go` file, but that may just make the logic there more confusing.
2. The current interface of passing in expected metrics would result in pretty verbose generated tests. We may need to add more helper functions that can aid in generating expected metrics, or just deal with verbose tests.

Question out of ignorance: Why do we need `generated_component_telemetry_test.go` for all of these different components? If it's always the same helper methods, why do we want to generate so much identical code rather than a central re-usable package? 

Maybe another part of this is simply that we don't want to have to rely on `componenttest` for `v1.0`?

Reply:
> Question out of ignorance: Why do we need generated_component_telemetry_test.go for all of these different components?

Totally valid question, I added it there as it would reduce the need for test package dependencies across component types. I didn't a processor to end up pulling in a dependency on receivers just to bring in a common test package. it would be possible to have these test separated per component type, but putting it in generated code seemed like a cheaper dependency.

Reply:
> The generated telemetry test from this PR would need to be moved up to each component's main directory, rather than internal/metadata

Would it be possible to move the tests code in generated_component_telemetry_test down into the internal package instead?

> The current interface of passing in expected metrics would result in pretty verbose generated tests. 

I would expect that we will eventually move to a similar model that we use for scrapers for "golden" metric files (i.e. https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/mongodbatlasreceiver/testdata/alerts/golden/metric-threshold-closed.yaml)

Reply:
> I didn't a processor to end up pulling in a dependency on receivers just to bring in a common test package. it would be possible to have these test separated per component type, but putting it in generated code seemed like a cheaper dependency.

Was there a reason that this was needed in the first place though, rather than depending on `componenttest`? From my understanding it doesn't seem to have the issue of pulling in different component type dependencies. It has caller methods that specific component types wouldn't use (`CheckExporterMetrics` wouldn't be used by a `receiver`), but that doesn't seem to be the issue you're referencing here.

_I realize this question is out of scope for these changes, just trying to understand context here._

> Would it be possible to move the tests code in generated_component_telemetry_test down into the internal package instead?

Yes, I like that solution much better, just wasn't sure if it would be too big of a breaking change 👍

Reply:
> Yes, I like that solution much better, just wasn't sure if it would be too big of a breaking change 👍

I don't think that's a huge change, would just require to make the funcs that are used by components public in the internal package, and update the references

Reply:
I gave the move an attempt, but we need to make a decision around the tradeoffs. I've filed https://github.com/open-telemetry/opentelemetry-collector/issues/11163 to discuss.